src/pretix/presale/checkoutflow.py

```diff
@@ -605,10 +630,96 @@ def forms(self):
                 formset.append(formsetentry)
         return formset
 
+    @property
+    def cross_selling_applicable_categories(self):
+        cart = self.positions
+        return [
+            (c, products_qs, discount_info) for (c, products_qs, discount_info) in
+            (
+                (c, *c.cross_sell_visible(cart, self.request.sales_channel.identifier))
+                for c in self.request.event.categories.filter(cross_selling_mode__isnull=False)
+            )
+            if products_qs is not None
+        ]
+
+    @cached_property
+    def cross_selling_is_applicable(self):
+        return any(len(items) > 0 for (category, items) in self.cross_selling_data)
+
+    @cached_property
+    def cross_selling_data(self):
+        class DummyCategory:
+            def __init__(self, rule, subevent=None):
+                self.id = rule.id
+                self.name = rule.name + (f" ({subevent})" if subevent else "")
+                self.description = rule.description
+
+        categories = self.cross_selling_applicable_categories
+        if self.event.has_subevents:
+            subevents = set(pos.subevent for pos in self.positions)
+            result = (
+                (DummyCategory(category, subevent), self._items_for_cross_selling(subevent, items_qs, discount_info), f'subevent_{subevent.pk}_')
+                for (category, items_qs, discount_info) in categories
+                for subevent in subevents
+            )
+        else:
+            result = (
+                (category, self._items_for_cross_selling(None, items_qs, discount_info))
+                for (category, items_qs, discount_info) in categories
+            )
+        return [(category, items) for (category, items) in result if len(items) > 0]
+
+    def _items_for_cross_selling(self, subevent, items_qs, discount_info):
```

Comment:
Would it make sense to move this into a separate module outside of the checkout step with clear boundaries (like a get_grouped_items, but one layer higher)? That would at least make it more testable and maybe even more understandable if the boundary is clearer

Reply:
yep, lives in `pretix/base/services/cross_selling.py` now